### ISSUE TITLE
An underlying issue of service.SecureString?

### ISSUE THREAD

**Issue by zhongyi-zhang** [Azure/open-service-broker-azure#490]
Does it extra call [json.Marshal](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/service/secure_string.go#L19) / [json.Unmarshal](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/service/secure_string.go#L25)?

Though they are symmetric, if you try to convert a struct which contains a SecureString member to map[string]interface{} by something like [service.GetMapFromStruct](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/service/map_utils.go#L38), the SecureString member would have an unexpected value. And it needs separate json.Marshal and json.Unmarshal to get the correct value.

I can show some example code: https://play.golang.org/p/fWTCTXCw81P. It is a SecureString type without your crypto.

**Comment by krancour** [comment]
Hey @zhongyi-zhang. Thanks for reporting this.

I don't think there's actually an issue here. Let me try to explain some of the more nuanced aspects of how marshaling/unmarshaling work in Go and why the code is the way it is. I can also explain why your example code exhibits the strange behavior you are observing.

First, [this line](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/service/secure_string.go#L19) might look odd. (Or maybe not?) I have bytes already and I'm marshaling those to... bytes. Let me explain that and I think it sets us up for understanding the rest of this...

First, the bytes that you get after encrypting are not necessarily ASCII bytes. It's a bunch of binary gibberish. To successfully include this information in JSON, it needs to be base64 encoded. As it so happens, when you marshal []byte in Go, what you get back is the original bytes base64 encoded then double quoted.

By way of example, let's assume you start with `SecureString("abc")` and no-op encryption. During the marshaling process, this becomes bytes `[97 98 99]` and gets base64 encoded as `[89 87 74 106]`, after the quotes are added to the beginning and end, you end up with `[34 89 87 74 106 34]`.

This ends up in the JSON doc looking like this:

```
{
  "foo": "YWJj"
}
```

Reversing the above requires stripping the quotes then base64 decoding.

In both directions, we rely on the `encoding/json` package to do all the heavy lifting. (Could have written this code to use `encoding/base64` + some ugly slice manipulation instead, but I thought it was more comprehensible as currently written.

I hope this makes sense and explains why the custom marshaling / unmarshaling code in `SecureString` is correct.

Now I can explain why you're encountering difficulty (as in your example) _despite_ `SecureString` code being correct.

In your example, everything works as you might expect all the way up through marshaling into the `tmp` variable. It's when you unmarshal that things go wrong for you. If you were unmarshaling into a struct of your type `a`, Go's unmarshaler would see that field `X` is of type `SecureString` and as a result, the custom unmarshaler for `SecureString` would be invoked at the proper point and everything would work as you expect. However, this is not what you are doing...

You are unmarshaling into a `map[string]interface{}`. When you do that, Go's unmarshaler doesn't get any hints about the type of each field in the JSON object. Unquoted `true` and `false` are treated as `bool`, anything else that is quoted is treated as a `string`, and anything else unquoted is treated as the numeric Go type most capable of holding any numeric value (integer or floating point)-- `float64`...

So-- when you unmarshal the JSON bytes in the `tmp` variable into `m map[string]interface{}`, `"YWJj"` is treated as a plain string. The custom unmarshaling code for `SecureString` is never invoked.

When you proceed, toward the end of your example to unmarshal bytes directly into `y SecureString`, Go's unmarshaler gets the hint and will properly invoke custom unmarshaling code for `SecureString`.

So... to summarize, the behavior you're seeing is normal Go behavior when you unmarshal JSON into a `map[string]interface{}` and the unmarshaler doesn't receive any hints about the types of fields. Custom unmarshaling code gets bypassed.

What this means, practically speaking, is that you probably shouldn't do this!

> if you try to convert a struct which contains a SecureString member to map[string]interface{} by something like service.GetMapFromStruct, the SecureString member would have an unexpected value.

That doesn't surprise me because I don't believe those struct <--> map conversions utilize the custom marshalers / unmarshalers. You will actually notice that after we introduced the `SecureString` type, we actually _stopped_ using the map utility functions in the (stable) modules. (They only still exist because I think they were useful somewhere in some bit of test code. Some experimental modules still use them, but note that experimental modules are currently excluded from the build. There's a lot of cleanup that we need to do on those modules to get them re-incorporated into the build.)

Sorry... I know this has been a long and complex explanation, but have I convinced you that nothing is wrong here?
